### PR TITLE
Closure conversion = coyoneda blog post

### PR DESCRIPTION
**Pull request by maxsnew**
Finally fulfilling my bet to @gasche

### CODE REVIEW

File: blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+real maps \\( f : A \to B \\) in \\( C\\) to get a map \\( q \circ f :
+A \to Q \\) and this behaves as expected with respect to identity and
+composition.  Then the Yoneda lemma says that any object is a
+"generalized object" by taking the real maps into it and that this
+interpretation doesn't add any new maps between the original objects.
+
+So the informal explanation of the coyoneda lemma is that if we have a
+\\( \exists B. (A \to B) \times Q(B) \\) then we can get a \\( f : A
+\to B \\) and a "\\(q : B \to Q\\)" and we can't inspect \\(B \\) in
+any way, then all we can really do is to compose the two maps.
```
Comment (bennn) [inline blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md]: "then all ...." makes this sentence hard to parse

Maybe say ".... if we have a `\exists` .... can get a `f` ... and a `q`. Furthermore we can compose the 2 maps but we cannot inspect `B` in any way"

File: blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+
+But there's a gap from here to applying this to a programming
```
Comment (bennn) [inline blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md]: Start a new section here?

File: blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+language, \\(Q(A) \\) and \\(\exists B. (A \to B) \times Q(B) \\) as
+described so far are *sets*, but we wanted an isomorphism of *types*
```
Comment (bennn) [inline blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md]: Suggestion: ".... *sets*. We have a map from a category C to the category of sets, but we wanted an isomorphism of *types* ...."

File: blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+the word "set" with "object of \\(V \\)".
+Specifically, we say a \\(V \\)-enriched category still has a set of
+objects \\(Ob \\), but for each pair of objects \\(A,B \in Ob \\) we
+get a \\( V\\)-object \\(\Hom(A,B) \\) of morphisms from \\(A \\) to
+\\( B \\).
+If \\(V \\) is a closed category, we can see \\(V \\) *itself* as a
+\\(V\\)-enriched category with the same objects and just making
+\\(\Hom(A,B) = A \to B \\).
+
+Then we can reinterpret the coyoneda lemma above by saying \\(V\\)
```
Comment (bennn) [inline blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md]: "`\\(V\\)`-enriched category" ?

File: blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+category instead of category and \\(V\\)-valued presheaf instead of
+presheaf. If we fix the \\(V\\)-category to be \\(V \\) itself, then a
```
Comment (bennn) [inline blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md]: same as above, I'm not sure what a V-category is

File: blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+The coYoneda lemma is a generalization of the equivalence described
+above.
+I'll start with the ordinary version which uses *coends* and
+*presheaves*.
+
+The coYoneda lemma says that for any category \\( C \\), *presheaf*,
+i.e. a functor \\( Q : C^{op} \to \Set \\), and object \\(A \in C \\),
+\\(Q(A) \\) is isomorphic to:
+\\[ \exists B. (A \to B) \times Q(B) \\]
+usually coends use an integral sign but I think \\(\exists \\) is a
```
Comment (bennn) [inline blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md]: After reading the part below about "Haskelletons" and the proof going through with existentials, I realized
(1) I really don't know what a coend is
(2) I don't know how coYoneda depends on coends

File: blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+equivalence, so things like as above:
+
+```
+let x = 3 in λ y. x + y ≡ λ y. 3 + y
+```
+
+To prove this we'll show that in a language with existential types the
+types `∃ Γ. Γ × (Γ × A -> B)` and `A \to B` are isomorphic.
+The usual proof is by parametricity, instead we'll use a closely
+related category-theoretic argument: the coYoneda lemma.
```
[the remark this answers is not in the document]
Comment (maxsnew) [reply blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md]: Nlab says "co-Yoneda" and I don't like hyphens so I'm calling it "coYoneda" except in title case where it's "CoYoneda".
Made it consistent in the next vsn

**Comment by maxsnew** [comment]
I've updated a lot based  on @bennn 's comments, adding  description of what coends are and changing the explanation of presheaves.

If everything is good, I'll push it on Monday

File: blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+M(B,B) \\) such that the \\(A, B\\) positions are related by some
+function \\(f : A \to B \\), then \\(pack_A(mA) = pack_B(mB)\\).
+More formally, this means that if you have a \\(m' : M(B,A) \\), then
+
+\\[ pack_B(M(B,f)(m')) = pack_A(M(f,A)(m'))\\]
+or in a point-free style:
+\\[ pack_B \circ M(B,f) = pack_A \circ M(f,A) : M(B,A) \to \exists X. M(X,X) \\]
+
+A function parameterized by types like \\(pack \\) that has this
+property is called a "co-wedge" from \\(M \\).
```
Comment (bennn) [inline blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md]: `*co-wedge*`

File: blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+
+A coend is a type \\(\exists M(X,X) \\) and a co-wedge \\(\forall
```
Comment (bennn) [inline blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md]: `\exists X. M(X,X)` ?

File: blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+A. pack_A : M(A,A) \to \exists X. M(X,X) \\) that are *universal*,
+i.e. any other family of maps \\(f_A : M(A,A) \to C\\) factors through
+\\(pack \\). This gives us the sytnax for existential elimination.
```
Comment (bennn) [inline blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md]: `pack_A` ?

File: blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+If you are familiar with parametricity, it is a good exercise to see
+why the usual condition for invariance wrt all *relations* implies
+that a parametric \\(pack, \exists X. M(X,X) \\) will form a cowedge.
+It seems that in general it would not be a universal co-wedge because
+a parametric exists is invariant under all relations and there are
+many relations that don't act like functions.
+
+### Presheaves
+
+Next, a presheaf is just another word for functor \\( Q : C^{op} \to
```
Comment (bennn) [inline blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md]: replace "another word for" with "a"

File: blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+\Set \\).
+Think of this as a set that is parameterised by a type of "inputs", so
+if you have a map in \\(C, f : A \to B\\) you get a function \\(Q(f) :
+Q(B) \to Q(A) \\) that "preprocesses" the inputs using \\(f
+\\). Functoriality ensures that preprocessing with the identity is
+just the identity and that composition or preprocessers is the
```
Comment (bennn) [inline blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md]: "composition of"

File: blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+So the informal explanation of the coYoneda lemma is that for any
+presheaf \\(Q \\), if we have an \\( \exists X. (A \to X) \times Q(X)
+\\), then since we can't inspect the \\(X \\) in any way, all we can
+really do is compose the \\(Q(X) \\) with the preprocesser from the
+function \\(A \to X \\), giving us a \\(Q(A) \\).
+
+### Enriched Categories and Enriched CoYoneda
+But there's a gap from here to applying this to a programming
+language, the coYoneda lemma as presented says that \\(Q(A) \\) and
+\\(\exists B. (A \to B) \times Q(B) \\) isomorphic as *sets*, but we
```
Comment (bennn) [inline blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md]: "are isomorphic"

File: blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md [hunk 14]
```diff
@@ -0,0 +1,294 @@
...
+*enriched* coYoneda lemma.  Let \\(V \\) be a category, then if \\( V
+\\) is sufficiently like the category of sets, then we can do a lot of
+category theory by replacing the word "set" with "object of \\(V \\)".
+
+Specifically, a \\(V \\)-enriched category (or just \\(V\\)-category)
+has a set of objects \\(Ob \\), but for each pair of objects \\(A,B
+\in Ob \\) we get a \\( V\\)-object \\(\Hom(A,B) \\) of morphisms from
+\\(A \\) to \\( B \\).  If \\(V \\) is a closed category, we can see
+\\(V \\) *itself* as a \\(V\\)-enriched category with the same objects
+and just making \\(\Hom(A,B) = A \to B \\).
```
Comment (bennn) [inline blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md]: `A \to B` here is weird to me

at first I thought this was a typo, using an arrow here
then I looked on nLab and see the notation `hom(A,B)` and `[A,B]`

suggestion: "we can see `\\(V \\)` *itself* as a V-category with the same objects because V contains an internal hom object corresponding to each `\Hom(A,B)`"

File: blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md [hunk 15]
```diff
@@ -0,0 +1,294 @@
...
+\\[ Q(\Delta) \cong \exists \Gamma. (\Delta \to \Gamma) \times Q(\Gamma)
+\\]
+
+Clearly based on the right hand side, \\(Q \\) should be \\( - \times
+A \to B \\) which gives us:
+
+\\[ \Delta \times A \to B \cong \exists \Gamma. (\Delta \to \Gamma) \times (\Gamma \times A \to B)
+\\]
+
+and if we pick \\(\Delta = 1\\), the unit type and the fact that \\(1
```
Comment (bennn) [inline blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md]: this sentence is too long,

suggestion: "Let `\Delta = 1`. Since `1 \times A \cong A` and `1 \to \Gamma \cong \Gamma` (at least ....), we get the desired result ...."

File: blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md [hunk 16]
```diff
@@ -0,0 +1,294 @@
...
+
+We want to prove the closure conversion isomorphism:
+
+\\[ A \to B \cong \exists \Gamma. \Gamma \times (\Gamma \times A \to B)
+\\]
+
+using the \\(V \\)-coYoneda lemma which says for any contravariant
+functor \\(Q : V^{op} \to V \\),
+
+\\[ Q(\Delta) \cong \exists \Gamma. (\Delta \to \Gamma) \times Q(\Gamma)
```
Comment (bennn) [inline blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md]: I think it'd be helpful to write `\forall Q \forall \Delta` here, then `\forall \Delta` in the next math block

File: blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md [hunk 17]
```diff
@@ -0,0 +1,294 @@
...
+\\]
+
+and if we pick \\(\Delta = 1\\), the unit type and the fact that \\(1
+\times A \cong A \\) and \\(1 \to \Gamma \cong \Gamma\\) (at least in
+a pure language) then we get the desired result by composition:
+
+\\[ A \to B \cong 1 \times A \to B \cong \exists \Gamma. (1 \to
+\Gamma) \times (\Gamma \times A \to B) \cong \exists \Gamma. \Gamma
+\times (\Gamma \times A \to B)\\]
+
```
Comment (bennn) [inline blog/_src/posts/2017-08-24-closure-conversion-coyoneda.md]: Add a "Conclusion" section?

Conclusion 1: Category theorists basically proved that closure conversion is equivalence preserving back in the 1970s
Conclusion 2: category theory is a good language for proving that "obvious" theorems are obviously true

**Comment by maxsnew** [comment]
@mhyee tweet it yo

**Comment by mhyee** [comment]
> @mhyee tweet it yo

I got your back, I powered up my time machine and went 10 minutes into the past: https://twitter.com/neu_prl/status/902177665930711041
